### ISSUE TITLE
Get styled text from RSyntaxTextArea bypassing clipboard

### ISSUE THREAD

**Issue by jblang** [bobbylight/RSyntaxTextArea#439]
**Describe the solution you'd like**
I want to be able to get a styled code snippet (retaining the syntax highlighting) from a RSyntaxTextArea and insert it into a JTextPane which contains other arbitrarily styled text.

**Are there any workarounds?**
I see that RSyntaxTextArea has a copyStyledText method, but I would rather transfer the text directly and not interfere with the user's clipboard.

**Comment by bobbylight** [comment]
Is the JTextPane using `HTMLEditorKit` or `RTFEditorKit` for its styled content?  If so then you should be able to use one of the utility methods that `copyAsStyledText()` uses under the covers:

```java
String html = HtmlUtil.getTextAsHtml(textArea, textArea.getSelectionStart(), textArea.getSelectionEnd());
// or
byte[] rtf = textArea.getTextAsRtf(textArea.getSelectionStart(), textArea.getSelectionEnd());
```

After typing that out, I'm not too sure that helps.  But what format would you want to get the content in other than `text/html` or `text/rtf`?

**Comment by jblang** [comment]
I'm not too familiar with the Java Text API so let me give you some context.  I'm using RSyntaxTextArea to replace a custom syntax highlighter in a [Logo interpreter](https://github.com/jblang/xlogo) that I inherited the code for. 

I have already replaced the text editor [with RSyntaxTextArea](https://github.com/jblang/xlogo/blob/syntax/src/xlogo/gui/Editor.java#L116-L135) and implemented a [LogoTokenMaker](https://github.com/jblang/xlogo/blob/syntax/src/xlogo/gui/LogoTokenMaker.java) to highlight Logo code:

![editor](https://user-images.githubusercontent.com/947942/158027599-0ba18f39-b91d-4a3a-9d0e-fe3872ed7909.png)

Syntax highlighting is also used in two other places: a command line for entering immediate commands, and a history panel that shows both the previously entered commands and the textual output from the logo programs. I have replaced the command line with a single-line [RSyntaxTextArea instance](https://github.com/jblang/xlogo/blob/syntax/src/xlogo/gui/Application.java#L59), but the history panel is trickier to replace since it displays not only logo code but also arbitrary output:

![main](https://user-images.githubusercontent.com/947942/158027372-56b0356c-f23a-4f91-96f9-f29482b17163.png)

The [HistoryPanel](https://github.com/jblang/xlogo/blob/syntax/src/xlogo/gui/HistoryPanel.java) class implements highligthing using a [HistoryTextPane](https://github.com/jblang/xlogo/blob/syntax/src/xlogo/gui/HistoryPanel.java#L124-L225) subclass of JTextPane. The HistoryTextPane uses a [HistoricLogoDocument](https://github.com/jblang/xlogo/blob/syntax/src/xlogo/document/HistoricLogoDocument.java), which is a subclass of a [LogoDocument](https://github.com/jblang/xlogo/blob/syntax/src/xlogo/document/LogoDocument.java), itself a subclass of DefaultStyledDocument. There used to be other subclasses of LogoDocument for the text editor and command line but I deleted them after replacing them with RSyntaxTextArea. From what I can tell the HistoricLogoDocument doesn't explicitly use either RTF or HTML internally but it does include the ability to export RTF from the history panel [using RTFEditorKit](https://github.com/jblang/xlogo/blob/syntax/src/xlogo/gui/HistoryPanel.java#L202-L215).

So basically what I want to do take the higlighted text from the command line RSyntaxTextArea and insert it (retaining the highlighting) into the HistoryPanel's document, along with whatever text might already be there from previous commands or output from the program.

**Comment by jblang** [comment]
I have simplified and consolidated the code since originally writing the previous comment. Now all of the style manipulation is consolidated in the [HistoryPanel](https://github.com/jblang/xlogo/blob/syntax/src/xlogo/gui/HistoryPanel.java) class. It has [various methods](https://github.com/jblang/xlogo/blob/syntax/src/xlogo/gui/HistoryPanel.java#L111-L147) which are called from the Logo code to set the color, font, size, and attributes of the text that is output by the print command.  The print command adds text to the history panel using the previously configured style via the [setText](https://github.com/jblang/xlogo/blob/syntax/src/xlogo/gui/HistoryPanel.java#L59) method. 

In addition to the output from the print command, the history panel has the history of commands that were executed from the command line.  I would still like to retain the syntax highlighting from the command line, but I could not figure out how to insert the text returned by RSyntaxTextArea.getTextAsRtf into the HistoryPanel's JTextPane.  For now I am simply adding commands to the history panel without any styling.

**Comment by jblang** [comment]
I've decided that having the program output mixed with the history of user commands is probably not the best UI anyway, so I will probably separate them in the future, which would render this request moot, so I'll go ahead and close it.  BTW, if you are interested in seeing how I'm using RSyntaxTextArea, I have just released [Beta 4](https://github.com/jblang/xlogo/releases/tag/1.0.0beta4) of XLogo which includes the replacement of the custom editor components with RSyntaxTextArea.
